packages/flutter/lib/src/material/navigation_drawer.dart

```diff
@@ -0,0 +1,416 @@
+/// a corresponding widget in the body of the [Scaffold].
+///
+/// ** See code in examples/api/lib/material/navigation_drawer/navigation_drawer.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [Scaffold.drawer], where one specifies a [Drawer] so that it can be
+///    shown.
+///  * [Scaffold.of], to obtain the current [ScaffoldState], which manages the
+///    display and animation of the drawer.
+///  * [ScaffoldState.openDrawer], which displays its [Drawer], if any.
+///  * <https://material.io/design/components/navigation-drawer.html>
+class NavigationDrawer extends StatelessWidget {
+  /// Creates a Material 3 Navigation Drawer component.
+  const NavigationDrawer({
+    super.key,
+    required this.children,
+    this.backgroundColor,
+    this.surfaceTintColor,
+    this.elevation,
+    this.onDestinationSelected,
+    this.selectedIndex = 0,
+  });
+
+  /// Sets the color of the [Material] that holds all of the [NavigationDrawer]'s
+  /// contents.
+  ///
+  /// If this is null, then [NavigationDrawerThemeData.backgroundColor] is used.
+  /// If that is also null, then it falls back to [Material]'s default.
+  final Color? backgroundColor;
+
+  /// Sets the color of the [Material] that holds all of the [NavigationDrawer]'s
+  /// contents.
+  ///
+  /// If this is null, then [NavigationDrawerThemeData.surfaceTintColor] is used.
+  /// If that is also null, then it falls back to [Material]'s default.
+  final Color? surfaceTintColor;
+
+  /// The elevation of the [NavigationDrawer] itself.
+  ///
+  /// If null, [NavigationDrawerThemeData.elevation] is used. If that
+  /// is also null, it will be 1.0.
+  final double? elevation;
+
+  /// Defines the appearance of the items within the navigation drawer.
+  ///
+  /// The list contains [NavigationDrawerDestination] widgets and/or customized
+  /// widgets like headlines and dividers.
+  final List<Widget> children;
+
+  /// The index into destinations for the current selected
+  /// [NavigationDrawerDestination] or null if no destination is selected.
+  ///
+  /// Only [NavigationDrawerDestination] is counted in index.
+  final int? selectedIndex;
+
+  /// Called when one of the destinations is selected.
+  ///
+  /// This callback usually updates the int passed to [selectedIndex].
+  ///
+  /// Upon updating [selectedIndex], the [NavigationDrawer] will be rebuilt.
+  final ValueChanged<int>? onDestinationSelected;
+
+  @override
+  Widget build(BuildContext context) {
+    final NavigationDrawerThemeData navigationDrawerTheme =
+        NavigationDrawerTheme.of(context);
+    final NavigationDrawerThemeData defaults =
+        _NavigationDrawerDefaultsM3(context);
+    final int totalNumberOfDestinations =
+        children.whereType<NavigationDrawerDestination>().toList().length;
+
+    final List<Widget> newChildren = <Widget>[];
+    for (int i = 0, j = 0; i < children.length; i++) {
+      final Widget e = children[i];
```

Comment:
This could be written in a more functional way with something like:

```dart
int destinationIndex = 0;
Widget _wrapChild(Widget child) {
  if (child is NavigationDrawerDestination) {
    return SelectableAnimatedBuilder(
            duration: const Duration(milliseconds: 500),
            isSelected: destinationIndex == selectedIndex,
            builder: (BuildContext context, Animation<double> animation) {
              return NavigationDestinationInfo(
                index: destinationIndex,
                // ...
        destinationIndex += 1;
    }
    return child;
}

final List<Widget> wrappedChildren = children.map(_wrapChild).toList();

Reply:
Done. Thanks!

Reply:
Actually after testing it I found out that adding " destinationIndex += 1" in "_wrapChild" will make it not working. Maybe because the function is run lazily?

So I will keep the "for" loop to make it work.